File: 2023/09_MirageMaintenance/report.py

# ======================================================================
# Mirage Maintenance
#   Advent of Code 2023 Day 09 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         r e p o r t . p y
# ======================================================================
"Report for the Advent of Code 2023 Day 09 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
from history import History

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                                 Report
# ======================================================================


class Report(object):   # pylint: disable=R0902, R0903, R0205
    "Object for Mirage Maintenance"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.histories = []

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self.histories = [History(text=line, part2=part2) for line in text]

    def next_values(self):
        "Return the sum of the next values for all of the histories"

        return sum(hist.next_value() for hist in self.histories)

    def prev_values(self):
        "Return the sum of the previous values for all of the histories"

        return sum(hist.prev_value() for hist in self.histories)

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------


if __name__ == '__main__':
    pass

# ======================================================================
# end                        r e p o r t . p y                       end
# ======================================================================

File: 2023/09_MirageMaintenance/report.go
// ======================================================================
// Mirage Maintenance
//   Advent of Code 2023 Day 09 -- Eric Wastl -- https://adventofcode.com
//
// go implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                         r e p o r t . g o
// ======================================================================
// A solver for the Advent of Code 2023 Day 09 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
)

// ----------------------------------------------------------------------
//
//	constants
//
// ----------------------------------------------------------------------
const ()

// ======================================================================
//                                                                 Report
// ======================================================================

// Report ... Object for Mirage Maintenance
type Report struct {
	Part2     bool
	Text      []string
	Histories []*History
}

// NewReport ... Create a Report object
func NewReport(part2 bool, text []string) (*Report, error) {

	// 1. Set the initial values
	p := new(Report)
	p.Part2 = part2
	p.Text = text

	// 2. Process text (if any)
	if len(p.Text) > 0 {
		err := p.processText(p.Text)
		if err != nil {
			return nil, err
		}
	}

	// 3. Return the new Report object
	return p, nil
}

// processText ... Assign values from text
func (p *Report) processText(text []string) error {

	// 1. Loop for each line of the text
	for indx, line := range text {

		// 2. Add the number to the entries
		history, err := NewHistory(p.Part2, line)
		if err != nil {
			return fmt.Errorf("error in line %d: %s", indx, err)
		}
		p.Histories = append(p.Histories, history)
	}

	// 3. Return success
	return nil
}

// PartOne ... Returns the solution for part one
func (p *Report) PartOne(verbose bool, limit int) string {

	// 1. Start with nothing
	result := 0

	// 2. Loop for all of the histories
	for _, history := range p.Histories {

		// 3. Accumulate the next values
		result = result + history.NextValue()
	}

	// 4. Return the solution for part one
	return fmt.Sprint(result)
}

// PartTwo ... Returns the solution for part two
func (p *Report) PartTwo(verbose bool, limit int) string {

	// 1. Start with nothing
	result := 0

	// 2. Loop for all of the histories
	for _, history := range p.Histories {

		// 3. Accumulate the next values
		result = result + history.PrevValue()
	}

	// 4. Return the solution for part two
	return fmt.Sprint(result)
}

// ======================================================================
// end                        r e p o r t . g o                       end
// ======================================================================
